src/intelligence_layer/learning/postgres_instruction_finetuning_data_repository.py

```diff
@@ -0,0 +1,101 @@
+from collections.abc import Iterable
+from itertools import islice
+from typing import Optional
+
+from sqlalchemy import ColumnElement, create_engine
+from sqlalchemy.orm import Query, sessionmaker
+
+from intelligence_layer.learning.instruction_finetuning_data_repository import (
+    InstructionFinetuningDataRepository,
+)
+from intelligence_layer.learning.models import (
+    Base,
+    InstructionFinetuningSample,
+    InstructionFinetuningSample_,
+)
+
+
+class PostgresInstructionFinetuningDataRepository(InstructionFinetuningDataRepository):
+    def __init__(
+        self,
+        database_url: str,
+    ) -> None:
+        self.engine = create_engine(database_url)
```

Comment:
Can we have connection pooling? https://docs.sqlalchemy.org/en/20/core/pooling.html#connection-pool-configuration if multiple users connect to the database it can reach a limit.

Reply:
Will do.

Reply:
I do not know if also make sense to use the postgresql+asyncpg. In order to not block the code, but maybe this is a topic for later.

Reply:
Yeah, sounds like for later. For now, I just introduced pooling as in the docs.